### ISSUE TITLE
failed: mustBePreparedToReceiveOnNextSignalsAfterHavingCalledSubscriptionCancel with new reactive-streams-tck

### ISSUE THREAD

**Issue by patriknw** [akka/akka#27033]
Trying upcoming reactive-streams-tck 1.0.3

See https://github.com/reactive-streams/reactive-streams-jvm/issues/453

One test failure, but it doesn't repeat deterministically. Seen it a few times in different tests several runs.

* akka.stream.tck.VirtualProcessorTest
* akka.stream.tck.TransformProcessorTest

```
[info] - required_spec208_mustBePreparedToReceiveOnNextSignalsAfterHavingCalledSubscriptionCancel *** FAILED ***
[info]   java.lang.AssertionError: Did not receive expected `request` call within 800 ms
[info]   at org.testng.Assert.fail(Assert.java:78)
[info]   at org.reactivestreams.tck.TestEnvironment.flopAndFail(TestEnvironment.java:292)
[info]   at org.reactivestreams.tck.TestEnvironment$Receptacle.next(TestEnvironment.java:1054)
[info]   at org.reactivestreams.tck.TestEnvironment$ManualPublisher.expectRequest(TestEnvironment.java:847)
[info]   at org.reactivestreams.tck.TestEnvironment$ManualPublisher.expectRequest(TestEnvironment.java:843)
[info]   at org.reactivestreams.tck.SubscriberWhiteboxVerification$6.run(SubscriberWhiteboxVerification.java:270)
[info]   at org.reactivestreams.tck.SubscriberWhiteboxVerification.subscriberTest(SubscriberWhiteboxVerification.java:504)
[info]   at org.reactivestreams.tck.SubscriberWhiteboxVerification.required_spec208_mustBePreparedToReceiveOnNextSignalsAfterHavingCalledSubscriptionCancel(SubscriberWhiteboxVerification.java:265)
[info]   at org.reactivestreams.tck.IdentityProcessorVerification.required_spec208_mustBePreparedToReceiveOnNextSignalsAfterHavingCalledSubscriptionCancel(IdentityProcessorVerification.java:622)
[info]   at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[info]   at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[info]   at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[info]   at java.lang.reflect.Method.invoke(Method.java:498)
[info]   at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:74)
[info]   at org.testng.internal.Invoker.invokeMethod(Invoker.java:673)
[info]   at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:846)
[info]   at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1170)
[info]   at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:125)
[info]   at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:109)
[info]   at org.testng.TestRunner.runWorkers(TestRunner.java:1147)
[info]   at org.testng.TestRunner.privateRun(TestRunner.java:749)
[info]   at org.testng.TestRunner.run(TestRunner.java:600)
[info]   at org.testng.SuiteRunner.runTest(SuiteRunner.java:317)
[info]   at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:312)
[info]   at org.testng.SuiteRunner.privateRun(SuiteRunner.java:274)
[info]   at org.testng.SuiteRunner.run(SuiteRunner.java:223)
[info]   at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
[info]   at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
[info]   at org.testng.TestNG.runSuitesSequentially(TestNG.java:1039)
[info]   at org.testng.TestNG.runSuitesLocally(TestNG.java:964)
[info]   at org.testng.TestNG.run(TestNG.java:900)
[info]   at org.scalatest.testng.TestNGSuiteLike.run(TestNGSuiteLike.scala:261)
[info]   at org.scalatest.testng.TestNGSuiteLike.run$(TestNGSuiteLike.scala:254)
[info]   at akka.stream.tck.AkkaIdentityProcessorVerification.run(AkkaIdentityProcessorVerification.scala:17)
[info]   at org.scalatest.testng.TestNGSuiteLike.runTestNG(TestNGSuiteLike.scala:248)
[info]   at org.scalatest.testng.TestNGSuiteLike.runTestNG$(TestNGSuiteLike.scala:227)
[info]   at akka.stream.tck.AkkaIdentityProcessorVerification.runTestNG(AkkaIdentityProcessorVerification.scala:17)
[info]   at org.scalatest.testng.TestNGSuiteLike.run(TestNGSuiteLike.scala:149)
[info]   at org.scalatest.testng.TestNGSuiteLike.run$(TestNGSuiteLike.scala:146)
[info]   at akka.stream.tck.AkkaIdentityProcessorVerification.run(AkkaIdentityProcessorVerification.scala:17)
[info]   at org.scalatest.tools.Framework.org$scalatest$tools$Framework$$runSuite(Framework.scala:317)
[info]   at org.scalatest.tools.Framework$ScalaTestTask.execute(Framework.scala:510)
[info]   at sbt.ForkMain$Run.lambda$runTest$1(ForkMain.java:304)
[info]   at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[info]   at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[info]   at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[info]   at java.lang.Thread.run(Thread.java:748)
[info]   Cause: java.lang.AssertionError: Did not receive expected `request` call within 800 ms
[info]   at org.testng.Assert.fail(Assert.java:89)
[info]   at org.reactivestreams.tck.TestEnvironment.flopAndFail(TestEnvironment.java:289)
[info]   at org.reactivestreams.tck.TestEnvironment$Receptacle.next(TestEnvironment.java:1054)
[info]   at org.reactivestreams.tck.TestEnvironment$ManualPublisher.expectRequest(TestEnvironment.java:847)
[info]   at org.reactivestreams.tck.TestEnvironment$ManualPublisher.expectRequest(TestEnvironment.java:843)
[info]   at org.reactivestreams.tck.SubscriberWhiteboxVerification$6.run(SubscriberWhiteboxVerification.java:270)
[info]   at org.reactivestreams.tck.SubscriberWhiteboxVerification.subscriberTest(SubscriberWhiteboxVerification.java:504)
[info]   at org.reactivestreams.tck.SubscriberWhiteboxVerification.required_spec208_mustBePreparedToReceiveOnNextSignalsAfterHavingCalledSubscriptionCancel(SubscriberWhiteboxVerification.java:265)
[info]   at org.reactivestreams.tck.IdentityProcessorVerification.required_spec208_mustBePreparedToReceiveOnNextSignalsAfterHavingCalledSubscriptionCancel(IdentityProcessorVerification.java:622)
[info]   at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[info]   at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[info]   at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[info]   at java.lang.reflect.Method.invoke(Method.java:498)
[info]   at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:74)
[info]   at org.testng.internal.Invoker.invokeMethod(Invoker.java:673)
[info]   at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:846)
[info]   at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1170)
[info]   at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:125)
[info]   at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:109)
[info]   at org.testng.TestRunner.runWorkers(TestRunner.java:1147)
[info]   at org.testng.TestRunner.privateRun(TestRunner.java:749)
[info]   at org.testng.TestRunner.run(TestRunner.java:600)
[info]   at org.testng.SuiteRunner.runTest(SuiteRunner.java:317)
[info]   at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:312)
[info]   at org.testng.SuiteRunner.privateRun(SuiteRunner.java:274)
[info]   at org.testng.SuiteRunner.run(SuiteRunner.java:223)
[info]   at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
[info]   at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
[info]   at org.testng.TestNG.runSuitesSequentially(TestNG.java:1039)
[info]   at org.testng.TestNG.runSuitesLocally(TestNG.java:964)
[info]   at org.testng.TestNG.run(TestNG.java:900)
[info]   at org.scalatest.testng.TestNGSuiteLike.run(TestNGSuiteLike.scala:261)
[info]   at org.scalatest.testng.TestNGSuiteLike.run$(TestNGSuiteLike.scala:254)
[info]   at akka.stream.tck.AkkaIdentityProcessorVerification.run(AkkaIdentityProcessorVerification.scala:17)
[info]   at org.scalatest.testng.TestNGSuiteLike.runTestNG(TestNGSuiteLike.scala:248)
[info]   at org.scalatest.testng.TestNGSuiteLike.runTestNG$(TestNGSuiteLike.scala:227)
[info]   at akka.stream.tck.AkkaIdentityProcessorVerification.runTestNG(AkkaIdentityProcessorVerification.scala:17)
[info]   at org.scalatest.testng.TestNGSuiteLike.run(TestNGSuiteLike.scala:149)
[info]   at org.scalatest.testng.TestNGSuiteLike.run$(TestNGSuiteLike.scala:146)
[info]   at akka.stream.tck.AkkaIdentityProcessorVerification.run(AkkaIdentityProcessorVerification.scala:17)
[info]   at org.scalatest.tools.Framework.org$scalatest$tools$Framework$$runSuite(Framework.scala:317)
[info]   at org.scalatest.tools.Framework$ScalaTestTask.execute(Framework.scala:510)
[info]   at sbt.ForkMain$Run.lambda$runTest$1(ForkMain.java:304)
[info]   at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[info]   at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[info]   at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[info]   at java.lang.Thread.run(Thread.java:748)
```

cc @viktorklang

**Comment by viktorklang** [comment]
Have a look at the TCK changes for milestone 1.0.3: https://github.com/reactive-streams/reactive-streams-jvm/issues?q=is%3Aclosed+milestone%3A1.0.3

**Comment by patriknw** [comment]
Looks like a new `defaultPollTimeoutMillis` was added. When not defined it's using the `defaultTimeoutMillis` for that parameter. After changing and specifying that parameter to 20 millis I can't reproduce the failure.
https://github.com/reactive-streams/reactive-streams-jvm/pull/452/files#diff-9c028d0d9c59f46577303e145146350bR79

We have:
```
  def publisherShutdownTimeoutMillis: Int = 3000
  def defaultTimeoutMillis: Int = 800
  def defaultNoSignalsTimeoutMillis: Int = 200
  def defaultPollTimeoutMillis: Int = 20
```

**Comment by patriknw** [comment]
I spoke to early. Still a problem:
```
[info] - required_spec208_mustBePreparedToReceiveOnNextSignalsAfterHavingCalledSubscriptionCancel *** FAILED ***
[info]   java.lang.AssertionError: Did not receive expected `request` call within 800 ms
[info]   at org.testng.Assert.fail(Assert.java:78)
[info]   at org.reactivestreams.tck.TestEnvironment.flopAndFail(TestEnvironment.java:292)
[info]   at org.reactivestreams.tck.TestEnvironment$Receptacle.next(TestEnvironment.java:1054)
[info]   at org.reactivestreams.tck.TestEnvironment$ManualPublisher.expectRequest(TestEnvironment.java:847)
[info]   at org.reactivestreams.tck.TestEnvironment$ManualPublisher.expectRequest(TestEnvironment.java:843)
[info]   at org.reactivestreams.tck.SubscriberWhiteboxVerification$6.run(SubscriberWhiteboxVerification.java:270)
[info]   at org.reactivestreams.tck.SubscriberWhiteboxVerification.subscriberTest(SubscriberWhiteboxVerification.java:504)
[info]   at org.reactivestreams.tck.SubscriberWhiteboxVerification.required_spec208_mustBePreparedToReceiveOnNextSignalsAfterHavingCalledSubscriptionCancel(SubscriberWhiteboxVerification.java:265)
[info]   at org.reactivestreams.tck.IdentityProcessorVerification.required_spec208_mustBePreparedToReceiveOnNextSignalsAfterHavingCalledSubscriptionCancel(IdentityProcessorVerification.java:622)
```

**Comment by patriknw** [comment]
The change to the TCK is here: https://github.com/reactive-streams/reactive-streams-jvm/pull/441#discussion-diff-234568341R270

@jroper Do you know more about this change and can give any hints on why Akka tests fail with upcoming TCK?

**Comment by jroper** [comment]
That changed was introduced due to this TCK verification failure that I found:

https://github.com/akka/akka/issues/25927

If `required_spec208_mustBePreparedToReceiveOnNextSignalsAfterHavingCall` is failing, then I guess the reason must be that Akka streams is allowing `signalCancel` to overtake the `signalRequest(1)`. In which case, Akka probably isn't implementing `signalRequest` correctly, since that requires that if called, that a request is made on the publisher, and if the cancel overtakes the request, then that's not happening.

A discussion about this was had [in the PR](https://github.com/reactive-streams/reactive-streams-jvm/pull/441/files#r234569272) between @viktorklang and I.

So solutions to this could be:

* Redefine the contract for `SubscriberPuppet.signalRequest` to specifically allow `signalCancel` to overtake `signalRequest`, and then modify `required_spec208_mustBePreparedToReceiveOnNextSignalsAfterHavingCall` to either not expect the request signal (but it must allow it), or to expect the request signal before it signals cancel. Or perhaps a second test can be added that tests either case.
* Modify Akka streams TCK puppet implementation such that when the puppet asks it to signal a request, it always signals a request.

**Comment by jrudolph** [comment]
Seemed to have failed before https://github.com/akka/akka/issues/22995 but then not anymore for a long time. So, maybe it is a timing issue that has been in here for a long while?

**Comment by jrudolph** [comment]
> If `required_spec208_mustBePreparedToReceiveOnNextSignalsAfterHavingCall` is failing, then I guess the reason must be that Akka streams is allowing `signalCancel` to overtake the `signalRequest(1)`. In which case, Akka probably isn't implementing `signalRequest` correctly, since that requires that if called, that a request is made on the publisher, and if the cancel overtakes the request, then that's not happening.

It's the `SubscriberPuppet` implementation by the TCK's `IdentityProcessorVerification`, so not sure how that could be fixed in Akka itself.

**Comment by jrudolph** [comment]
> Seemed to have failed before in #22995 but then not anymore for a long time. So, maybe it is a timing issue that has been in here for a long while?

That is, before it failed on `signalNext` now it fails on the new assertion `expectRequest`, so it may or may not be related.

**Comment by jrudolph** [comment]
> > If `required_spec208_mustBePreparedToReceiveOnNextSignalsAfterHavingCall` is failing, then I guess the reason must be that Akka streams is allowing `signalCancel` to overtake the `signalRequest(1)`. In which case, Akka probably isn't implementing `signalRequest` correctly, since that requires that if called, that a request is made on the publisher, and if the cancel overtakes the request, then that's not happening.
> 
> It's the `SubscriberPuppet` implementation by the TCK's `IdentityProcessorVerification`, so not sure how that could be fixed in Akka itself.

Following up on this, see https://github.com/reactive-streams/reactive-streams-jvm/pull/441/files#r288028712:

> These tests are also used in the IdentityProcessorVerification tests which implements the SubscriberPuppet as forwarding the triggers as is to the processor under test. So, IIUC the invariants you state would have to be valid for any identity processor implementation you might want to test. That would mean, that you cannot use the TCK to check processor implementations that do coalescing of requests and cancel.

I haven't checked where akka-stream would do the coalescing of request and cancel but in the case of `VirtualProcessorTest` there are a lot of complex stages and akka-stream implementation details involved, so I'd guess it's possible to happen.

**Comment by jroper** [comment]
Addressed here: https://github.com/reactive-streams/reactive-streams-jvm/pull/462

**Comment by johanandren** [comment]
We are up on 1.0.3-RC1 (#27513) so this was possibly fixed byt that but lets leave it open for some while longer to se if the failure pops up again.

**Comment by viktorklang** [comment]
1.0.3 is released so it is possible to bump akka’s version of it 👍
-- 
Cheers,
√

**Comment by patriknw** [comment]
@johanandren This is supposed to be fixed by 1.0.3 (and RC1) so I think we can update and close this. Create new ticket if it fails again.

**Comment by johanandren** [comment]
Alright, I'll PR update to 1.0.3 final and close this.
